subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/execution/ResidualProgramCompiler.kt

```diff
@@ -0,0 +1,754 @@
+        ALOAD(Vars.PluginRequestCollector)
+        LDC(plugins.fragment.lineNumber)
+        INVOKEVIRTUAL(
+            pluginRequestCollectorType,
+            "createSpec",
+            "(I)Lorg/gradle/plugin/use/PluginDependenciesSpec;")
+    }
+
+    private
+    val pluginRequestCollectorType by unsafeLazy { PluginRequestCollector::class.internalName }
+
+    private
+    fun MethodVisitor.invokeApplyPluginsTo() {
+        invokeHost(
+            "applyPluginsTo",
+            "(Lorg/gradle/kotlin/dsl/support/KotlinScriptHost;Lorg/gradle/plugin/management/internal/PluginRequests;)V")
+    }
+
+    private
+    fun ClassWriter.overrideLoadSecondStageFor(sourceFilePath: String, originalPath: String) {
+        publicMethod(
+            "loadSecondStageFor",
+            "(Lorg/gradle/kotlin/dsl/execution/ExecutableProgram\$Host;Lorg/gradle/kotlin/dsl/support/KotlinScriptHost;Ljava/lang/String;Lorg/gradle/internal/hash/HashCode;)Ljava/lang/Class;",
+            "(Lorg/gradle/kotlin/dsl/execution/ExecutableProgram\$Host;Lorg/gradle/kotlin/dsl/support/KotlinScriptHost<*>;Ljava/lang/String;Lorg/gradle/internal/hash/HashCode;)Ljava/lang/Class<*>;"
+        ) {
+
+            emitCompileSecondStageScript(sourceFilePath, originalPath)
+            ARETURN()
+        }
+    }
+
+    private
+    fun MethodVisitor.emitEvaluateSecondStageOf() {
+        // programHost.evaluateSecondStageOf(...)
+        ALOAD(Vars.ProgramHost)
+        ALOAD(Vars.Program)
+        ALOAD(Vars.ScriptHost)
+        LDC(programTarget.name + "/" + programKind.name + "/stage2")
+        // Move HashCode value to a static field so it's cached across invocations
+        loadHashCode(originalSourceHash)
+        invokeHost(
+            ExecutableProgram.Host::evaluateSecondStageOf.name,
+            "(Lorg/gradle/kotlin/dsl/execution/ExecutableProgram\$StagedProgram;Lorg/gradle/kotlin/dsl/support/KotlinScriptHost;Ljava/lang/String;Lorg/gradle/internal/hash/HashCode;)V")
+    }
+
+    private
+    fun MethodVisitor.emitCompileSecondStageScript(sourceFilePath: String, originalScriptPath: String) {
+        ALOAD(Vars.ProgramHost)
+        LDC(sourceFilePath)
+        LDC(originalScriptPath)
+        ALOAD(Vars.ScriptHost)
+        ALOAD(3)
+        ALOAD(4)
+        GETSTATIC(programKind)
+        GETSTATIC(programTarget)
+        invokeHost(
+            ExecutableProgram.Host::compileSecondStageScript.name,
+            "(" +
+                "Ljava/lang/String;Ljava/lang/String;" +
+                "Lorg/gradle/kotlin/dsl/support/KotlinScriptHost;" +
+                "Ljava/lang/String;Lorg/gradle/internal/hash/HashCode;" +
+                "Lorg/gradle/kotlin/dsl/execution/ProgramKind;" +
+                "Lorg/gradle/kotlin/dsl/execution/ProgramTarget;" +
+                ")Ljava/lang/Class;")
+    }
+
+    private
+    fun ClassVisitor.overrideExecute(methodBody: MethodVisitor.() -> Unit) {
```

Comment:
So I'm seeing this called a bunch of places, but I have no idea what it's doing or why. Can you explain what this is doing and why?